File: Intermediate/21.Detect and remove a cycle in a linked list/Solution.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

void printList(struct Node* head) {
    struct Node* temp = head;
    while (temp) {
        printf("%d -> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

int detectAndRemoveCycle(struct Node* head) {
    struct Node *slow = head, *fast = head;

    // Detect Cycle
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) {
            break;
        }
    }

    if (!fast || !fast->next) {
        return 0; // No cycle
    }

    // Remove Cycle
    slow = head;
    while (slow->next != fast->next) {
        slow = slow->next;
        fast = fast->next;
    }

    fast->next = NULL; // Break the cycle
    return 1;
}

int main() {
    int n, value, cyclePosition;

    printf("Enter the number of nodes: ");
    scanf("%d", &n);

    struct Node *head = NULL, *tail = NULL, *cycleNode = NULL;
    for (int i = 0; i < n; i++) {
        printf("Enter value for node %d: ", i + 1);
        scanf("%d", &value);

        struct Node* newNode = createNode(value);
        if (!head) {
            head = tail = newNode;
        } else {
            tail->next = newNode;
            tail = newNode;
        }

        if (i == cyclePosition - 1) {
            cycleNode = newNode;
        }
    }

    printf("Enter the position to create a cycle (0 for no cycle): ");
    scanf("%d", &cyclePosition);
    if (cyclePosition > 0) {
        tail->next = cycleNode;
    }

    if (detectAndRemoveCycle(head)) {
        printf("Cycle detected and removed.\n");
    } else {
        printf("No cycle detected.\n");
    }

    printf("Updated List: ");
    printList(head);

    return 0;
}

File: Intermediate/21.Detect and remove a cycle in a linked list/Solution.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def detect_and_remove_cycle(head):
    slow, fast = head, head

    # Detect Cycle
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            break

    if not fast or not fast.next:
        return False  # No cycle

    # Remove Cycle
    slow = head
    while slow.next != fast.next:
        slow = slow.next
        fast = fast.next

    fast.next = None  # Break the cycle
    return True

def print_list(head):
    current = head
    while current:
        print(current.data, end=" -> ")
        current = current.next
    print("NULL")

if __name__ == "__main__":
    n = int(input("Enter the number of nodes: "))
    head = None
    tail = None
    nodes = []

    for i in range(n):
        value = int(input(f"Enter value for node {i + 1}: "))
        new_node = Node(value)
        if not head:
            head = tail = new_node
        else:
            tail.next = new_node
            tail = new_node
        nodes.append(new_node)

    cycle_position = int(input("Enter the position to create a cycle (0 for no cycle): "))
    if cycle_position > 0:
        tail.next = nodes[cycle_position - 1]

    if detect_and_remove_cycle(head):
        print("Cycle detected and removed.")
    else:
        print("No cycle detected.")

    print("Updated List:")
    print_list(head)
